File: core/parser/expression.ts
import { Node, NodeParser } from "./node.ts";
import { WhistleParser } from "./parser.ts";
import {
  Literal,
  ParseBooleanLiteral,
  ParseIntegerLiteral,
  ParseFloatLiteral,
  ParseStringLiteral,
  ParseCharacterLiteral,
  ParseNoneLiteral,
} from "./literal.ts";
import {
  UnaryOperator,
  BinaryOperator,
  IsUnaryOperator,
  ParseUnaryOperator,
  IsBinaryOperator,
  GetBinaryOperator,
  ParseBinaryOperator,
} from "./operator.ts";

export type Expression =
  | UnaryExpression
  | BinaryExpression
  | IfExpression
  | FunctionCall
  | VariableAccess
  | Grouping
  | Literal;

export const ParseExpression: NodeParser<Expression> = (
  parser: WhistleParser,
) => {
  return ParseExpressionWithPrecedence(parser, -1);
};

export const ParseExpressionWithPrecedence: NodeParser<Expression> = (
  parser: WhistleParser,
  precedence: number,
) => {
  let left: PrimaryExpression | UnaryExpression | BinaryExpression =
    ParsePrimaryOrUnaryExpression(parser);

  while (
    IsBinaryOperator(parser.current) &&
    GetBinaryOperator(parser.current).precedence > precedence
  ) {
    left = ParseBinaryExpression(parser, left);
  }

  return left;
};

export const ParsePrimaryOrUnaryExpression: NodeParser<
  PrimaryExpression | UnaryExpression
> = (parser: WhistleParser) => {
  if (IsUnaryOperator(parser.current)) {
    return ParseUnaryExpression(parser);
  } else {
    return ParsePrimaryExpression(parser);
  }
};

export interface UnaryExpression extends Node<{
  operator: UnaryOperator;
  operand: Expression;
}> {
  type: "UnaryExpression";
}

export const ParseUnaryExpression: NodeParser<UnaryExpression> = (
  parser: WhistleParser,
) => {
  return {
    type: "UnaryExpression",
    value: {
      operator: ParseUnaryOperator(parser),
      operand: ParsePrimaryOrUnaryExpression(parser),
    },
  };
};

export interface BinaryExpression extends Node<{
  operandLeft: Expression;
  operator: BinaryOperator;
  operandRight: Expression;
}> {
  type: "BinaryExpression";
}

export const ParseBinaryExpression: NodeParser<BinaryExpression> = (
  parser: WhistleParser,
  left: Expression,
) => {
  const operator = ParseBinaryOperator(parser);

  return {
    type: "BinaryExpression",
    value: {
      operandLeft: left,
      operator,
      operandRight: ParseExpressionWithPrecedence(parser, operator.precedence),
    },
  };
};

export type PrimaryExpression =
  | Literal
  | FunctionCall
  | VariableAccess
  | Grouping;

export const ParsePrimaryExpression: NodeParser<PrimaryExpression> = (
  parser: WhistleParser,
) => {
  switch (parser.current.type) {
    case "Boolean":
      return ParseBooleanLiteral(parser);
    case "Integer":
      return ParseIntegerLiteral(parser);
    case "Float":
      return ParseFloatLiteral(parser);
    case "Character":
      return ParseCharacterLiteral(parser);
    case "String":
      return ParseStringLiteral(parser);
    case "Type":
      if (parser.is({ type: "Type", value: "none" })) {
        return ParseNoneLiteral(parser);
      } else {
        throw `Could not parse type "${parser.current}" as expression`;
      }
    case "Identifier":
      if (parser.is({ type: "LeftParenthesis" }, parser.next)) {
        return ParseFunctionCall(parser);
      } else {
        return ParseVariableAccess(parser);
      }
    case "LeftParenthesis":
      return ParseGrouping(parser);
  }

  throw `Could not parse primary expression ${JSON.stringify(
    parser.current,
  )}`;
};

export interface IfExpression extends Node<{
  condition: Expression;
  then: Expression;
  else: Expression;
}> {
  type: "IfExpression";
}

export const ParseIfExpression: NodeParser<IfExpression> = (
  parser: WhistleParser,
) => {
  parser.eat({ type: "Keyword", value: "if" });

  return {
    type: "IfExpression",
    value: {
      condition: ParseExpression(parser),
      then: ParseExpression(parser),
      else: parser.eat({ type: "Keyword", value: "else" }) &&
        ParseExpression(parser),
    },
  };
};

export interface FunctionCall extends Node<{
  name: string;
  parameters: Expression[];
}> {
  type: "FunctionCall";
}

export const ParseFunctionCall: NodeParser<FunctionCall> = (
  parser: WhistleParser,
) => {
  return {
    type: "FunctionCall",
    value: {
      name: parser.eat({ type: "Identifier" }).value,
      parameters: parser.delimited(
        { type: "LeftParenthesis", value: "(" },
        { type: "RightParenthesis", value: ")" },
        { type: "Comma", value: "," },
        () => ParseExpression(parser),
      ),
    },
  };
};

export interface VariableAccess extends Node<{
  name: string;
}> {
  type: "VariableAccess";
}

export const ParseVariableAccess: NodeParser<VariableAccess> = (
  parser: WhistleParser,
) => {
  return {
    type: "VariableAccess",
    value: {
      name: parser.eat({ type: "Identifier" }).value,
    },
  };
};

export interface Grouping extends Node<Expression> {
  type: "Grouping";
}

export const ParseGrouping: NodeParser<Grouping> = (parser: WhistleParser) => {
  parser.eat({ type: "LeftParenthesis", value: "(" });

  const expression = ParseExpression(parser);

  parser.eat({ type: "RightParenthesis", value: ")" });

  return {
    type: "Grouping",
    value: expression,
  };
};

File: core/src/parser/program/expression.rs
use crate::parser::parser::*;
use super::operator::*;
use crate::parser::ast::*;
use super::literal::*;
use super::ident::*;
pub use crate::lexer::*;

pub fn parse_expression(parser:&mut Parser) -> Option<Expr> {
    parse_expression_with_precedence(parser, -1)
}

pub fn parse_expression_with_precedence(parser:&mut Parser, precedence: i32) -> Option<Expr> {
    if let Some(expression) = parse_unary_or_primary_expression(parser) {
        let previous = Expr::Unary(expression);
        if let Some(Token::Operator(operator)) = parser.peek() {
            if Operator::is_binary(operator) {
                return parse_binary_expression(parser, previous)
            }
        } 
        return Some(previous)
    }
    None
}

pub fn parse_unary_or_primary_expression(parser:&mut Parser) -> Option<UnaryExpr> {
    if let Some(Token::Operator(operator)) = parser.peek(){
        if Operator::is_unary(operator) {
            return parse_unary_expression(parser)
        }
    }
    return parse_primary_expression(parser)
}

pub fn parse_unary_expression(parser:&mut Parser) -> Option<UnaryExpr> {
    if let Some(expression) = parse_unary_or_primary_expression(parser) {
        return Some(UnaryExpr::UnaryOp {
            op: parse_unary_operator(parser),
            expr: Box::new(expression)
        })
    }
    None
}

pub fn parse_binary_expression(parser:&mut Parser, previous: Expr) -> Option<Expr> {
    if let Some(expression) = parse_expression_with_precedence(parser , 0) { 
        return Some(Expr::Binary{
            lhs: Box::new(previous),
            op: parse_binary_operator(parser),
            //TODO: precedence
            rhs: Box::new(expression)
        })
    }
    None
}

pub fn parse_primary_expression(parser:&mut Parser) -> Option<UnaryExpr> {
    if let Some(current) = parser.peek() {
        //TODO: ParseConditional, ParseSelector
        let primary = match current {
            Token::BoolLit(_current) => parse_boolean_literal(parser),
            Token::IntLit(_current) => parse_integer_literal(parser),
            Token::FloatLit(_current) => parse_float_literal(parser),
            Token::CharLit(_current) => parse_char_literal(parser),
            Token::StrLit(_current) => parse_string_literal(parser),
            Token::Ident(_current) => 
                if parser.is_tok_eq(Token::Punc(Punc::LeftParen), 1) {
                    parse_function_call(parser)
                } else {
                    parse_variable_access(parser)
                }
            Token::Punc(Punc::LeftParen) => parse_grouping(parser),
            _ => {println!("Could not parse expression {:?}", current); None}
        };
        if let Some(primary) = primary {
            return Some(UnaryExpr::Primary(primary))
        }
    }
    None
}

pub fn parse_function_call(parser:&mut Parser) -> Option<PrimaryExpr> {
    if let Some(name) = parse_ident(parser){
        if parser.eat_type(Token::Punc(Punc::LeftParen)).is_some() {
            let expressions = parser.until_is(|parser| {
                let expr = parse_expression(parser);
                parser.eat_type(Token::Punc(Punc::Comma));
                expr
            }, Token::Punc(Punc::RightBrace));
            
            if parser.eat_type(Token::Punc(Punc::RightParen)).is_some() {
                return Some(PrimaryExpr::Arguments {
                    prim: Box::new(PrimaryExpr::Operand(Operand::Ident(name.to_string()))),
                    args: expressions
                })
            }
        }
    }
    None
}

pub fn parse_variable_access(parser:&mut Parser) -> Option<PrimaryExpr> {
    if let Some(name) = parse_ident(parser) {
        return Some(PrimaryExpr::Operand(Operand::Ident(name.to_string())))
    }
    None
}

pub fn parse_grouping(parser:&mut Parser) -> Option<PrimaryExpr> {
    if parser.eat_type(Token::Punc(Punc::LeftParen)).is_some() {
        if let Some(expr) = parse_expression(parser){
            if parser.eat_type(Token::Punc(Punc::RightParen)).is_some() {
                return Some(PrimaryExpr::Operand(Operand::Grouping(Box::new(expr))))
            }
        }
    }
    None
}